packages/opentelemetry-api/src/api/global-utils.ts

```diff
@@ -15,54 +15,74 @@
-  [GLOBAL_TRACE_API_KEY]: Get<TracerProvider>;
-}>;
+export function registerGlobal(type: 'trace', instance: TracerProvider): void;
+export function registerGlobal(type: 'metrics', instance: MeterProvider): void;
+export function registerGlobal(type: 'context', instance: ContextManager): void;
+export function registerGlobal(
+  type: 'propagation',
+  instance: TextMapPropagator
+): void;
+export function registerGlobal(type: keyof OTelGlobalApi, instance: any) {
+  _global[GLOBAL_OPENTELEMETRY_API_KEY] =
+    _global[GLOBAL_OPENTELEMETRY_API_KEY] ?? {};
 
-export const _global = _globalThis as OtelGlobal;
+  const api = _global[GLOBAL_OPENTELEMETRY_API_KEY]!;
+  if (api[type]) {
+    // already registered an API of this type
+    return;
+  }
 
-/**
- * Make a function which accepts a version integer and returns the instance of an API if the version
- * is compatible, or a fallback version (usually NOOP) if it is not.
- *
- * @param requiredVersion Backwards compatibility version which is required to return the instance
- * @param instance Instance which should be returned if the required version is compatible
- * @param fallback Fallback instance, usually NOOP, which will be returned if the required version is not compatible
- */
-export function makeGetter<T>(
-  requiredVersion: number,
-  instance: T,
-  fallback: T
-): Get<T> {
-  return (version: number): T =>
-    version === requiredVersion ? instance : fallback;
+  api[type] = {
+    instance,
+    version: VERSION,
+  };
 }
 
-/**
- * A number which should be incremented each time a backwards incompatible
- * change is made to the API. This number is used when an API package
- * attempts to access the global API to ensure it is getting a compatible
- * version. If the global API is not compatible with the API package
- * attempting to get it, a NOOP API implementation will be returned.
- */
-export const API_BACKWARDS_COMPATIBILITY_VERSION = 3;
+export function getGlobal(type: 'trace'): Signal<TracerProvider> | undefined;
+export function getGlobal(type: 'metrics'): Signal<MeterProvider> | undefined;
+export function getGlobal(type: 'context'): Signal<ContextManager> | undefined;
+export function getGlobal(
+  type: 'propagation'
+): Signal<TextMapPropagator> | undefined;
+export function getGlobal(type: keyof OTelGlobalApi) {
+  return _global[GLOBAL_OPENTELEMETRY_API_KEY]?.[type];
+}
+
+export function unregisterGlobal(type: keyof OTelGlobalApi) {
+  const api = _global[GLOBAL_OPENTELEMETRY_API_KEY];
+
+  if (api) {
+    delete api[type];
+  }
+}
+
+export function isCompatible(version: string) {
+  return semver.satisfies(version, acceptableRange);
+}
+
+type OTelGlobal = Partial<{
+  [GLOBAL_OPENTELEMETRY_API_KEY]: OTelGlobalApi;
+}>;
+
+type OTelGlobalApi = Partial<{
+  trace: Signal<TracerProvider>;
+  metrics: Signal<MeterProvider>;
+  context: Signal<ContextManager>;
+  propagation: Signal<TextMapPropagator>;
+}>;
+
+type Signal<T> = {
+  instance: T;
+  version: string;
```

Comment:
I think the version should be part of `OTelGlobalApi` and the first user should allocate it. Otherwise it may happen that e.g. propagation is of version 0.1.0 and context of 1.2.3.

This would result in all users are unhappy because at least context, trace and propagation are tight coupled.

Reply:
Hmm this solves the typing issue i've been having too.